Database-Features/hiding-join-conditions.md

```diff
@@ -0,0 +1,123 @@
+# Hiding Join Conditions
+
+<!-- meta
+   LAST_VERIFIED: Exasol 8.33.0
+-->
+
+## Problem
+
+While Exasol's query optimizer is quite smart in some places, it can be stubborn in others.
+One such stubborn place is located in join index creation:
+
+For any join, the query engine will always request &mdash; and create, if necessary &mdash; an index containing all columns that are referred using equality-join conditions[^eqj]:
+
+[^eqj]: Equality operator, and one side of the operation contains only columns of the adjoined table.
+
+This can lead to the different problems as shown below.
+
+**Notes:**
+
+- All examples use the TPC-DS data model
+- All examples use `LEFT JOIN` to enforce a join order for demonstration purposes
+- The smart part of Exasol's query optimizer might 'construct' the situations below by transporting or copying given conditions from `HAVING` to `WHERE` clauses and from `WHERE` to the respective JOIN operations[^conversion].
+For example, even if all joins inside a view are well-formed and contain only the necessary columns, filter conditions provided when selecting from the view may change join columns.
+
+[^conversion]: Note that any `WHERE` condition on an inner table that does not handle null values will automatically convert the respective join to an *inner join*.
+
+### 1 - Extra Index Columns
+
+```sql
+SELECT count(*)
+FROM customer
+LEFT JOIN customer_address
+   ON c_current_addr_sk = ca_address_sk
+   AND c_birth_country = ca_country
+```
+
+Even though `ca_address_sk` is actually a primary key, the query engine wants to join on both columns to get maximum index selectivity:
+
+|PART_ID|PART_NAME|PART_INFO|OBJECT_NAME|REMARKS|
+|---|---|---|---|---|
+|1|COMPILE / EXECUTE|(null)|(null)|(null)|
+|2|INDEX CREATE|(null)|CUSTOMER_ADDRESS|GLOBAL INDEX (CA_ADDRESS_SK,CA_COUNTRY)|
+|3|**INDEX CREATE**|on REPLICATED table|CUSTOMER_ADDRESS|**LOCAL INDEX (CA_ADDRESS_SK,CA_COUNTRY)**|
+|4|SCAN|(null)|CUSTOMER|(null)|
+|5|OUTER JOIN|on REPLICATED table|CUSTOMER_ADDRESS|CUSTOMER(C_CURRENT_ADDR_SK,C_BIRTH_COUNTRY) **=> LOCAL INDEX (CA_ADDRESS_SK,CA_COUNTRY)**|
```

Comment:
```suggestion
|5|OUTER JOIN|on REPLICATED table|CUSTOMER_ADDRESS|CUSTOMER(C_CURRENT_ADDR_SK,C_BIRTH_COUNTRY) **=&gt; LOCAL INDEX (CA_ADDRESS_SK,CA_COUNTRY)**|
```

There are three more.

Do you want to change them or try if it works in the current form?

Reply:
changing them. Maybe also a candidate for a github action...